lib/tbot/service_bot_identity.go

```diff
@@ -271,14 +335,21 @@ func (s *identityService) Run(ctx context.Context) error {
 		service: s.String(),
 		name:    "bot-identity-renewal",
 		f: func(ctx context.Context) error {
-			return s.renew(ctx, storageDestination)
+			if err := s.renew(ctx, storageDestination); err != nil {
+				return err
+			}
+			s.unblockWaiters()
+			return nil
 		},
-		interval:             s.cfg.CredentialLifetime.RenewalInterval,
-		exitOnRetryExhausted: true,
-		retryLimit:           botIdentityRenewalRetryLimit,
-		log:                  s.log,
-		reloadCh:             reloadCh,
-		waitBeforeFirstRun:   true,
+		interval:   s.cfg.CredentialLifetime.RenewalInterval,
+		retryLimit: botIdentityRenewalRetryLimit,
```

Comment:
As I'm using the "normal" periodic renewal process to recover from failed initialization, it will give up after 7 attempts and wait for the next renewal interval (default: 20 minutes).

Instead, I wonder if we need a separate exponential backoff loop without the timer. What do you think @strideynet?

Reply:
I'm thinking something like this before the `runOnInterval`:


```Go
if !s.IsReady() {
	retry, err := retryutils.NewRetryV2(retryutils.RetryV2Config{
		Driver: retryutils.NewExponentialDriver(1 * time.Second),
		Max:    1 * time.Minute,
		Jitter: retryutils.HalfJitter,
	})
	if err != nil {
		return trace.Wrap(err, "creating retry")
	}

	for {
		retry.Inc()

		s.log.InfoContext(ctx, "Unable to renew bot identity on startup. Waiting to retry", "wait", retry.Duration())

		select {
		case <-retry.After():
		case <-ctx.Done():
			return nil
		}

		if err := s.renew(ctx, storageDestination); err == nil {
			s.unblockWaiters()
			break
		}
	}
}

```

Reply:
Yeah - that seems to make sense to me.